connector/src/main/java/org/geysermc/connector/entity/BoatEntity.java

```diff
@@ -26,28 +26,107 @@
+    // Looks too fast and too choppy with 0.1f, which is how I believe the Microsoftian client handles it
+    private final float ROWING_SPEED = 0.05f;
+
     public BoatEntity(long entityId, long geyserId, EntityType entityType, Vector3f position, Vector3f motion, Vector3f rotation) {
         super(entityId, geyserId, entityType, position.add(0d, entityType.getOffset(), 0d), motion, rotation);
     }
 
     @Override
     public void moveAbsolute(GeyserSession session, Vector3f position, Vector3f rotation, boolean isOnGround) {
+        // Rotation is basically only called when entering/exiting a boat.
         super.moveAbsolute(session, position.add(0d, this.entityType.getOffset(), 0d), rotation, isOnGround);
     }
 
     @Override
     public void updateBedrockMetadata(EntityMetadata entityMetadata, GeyserSession session) {
         if (entityMetadata.getId() == 10) {
             metadata.put(EntityData.VARIANT, (int) entityMetadata.getValue());
+        } else if (entityMetadata.getId() == 11) {
+            is_paddling_left = (boolean) entityMetadata.getValue();
+            if (!is_paddling_left) {
+                metadata.put(EntityData.PADDLE_TIME_LEFT, 0f);
+            }
+            else {
+                // Java sends simply "true" and "false" (is_paddling_left), Bedrock keeps sending packets as you're rowing
+                // This is an asynchronous method that emulates Bedrock rowing until "false" is sent.
+                paddle_time_left = 0f;
+                session.getConnector().getGeneralThreadPool().schedule(() ->
+                                updateLeftPaddle(session, entityMetadata),
+                        100,
+                        TimeUnit.MILLISECONDS
+                );
+            }
+        }
+        else if (entityMetadata.getId() == 12) {
+            is_paddling_right = (boolean) entityMetadata.getValue();
+            if (!is_paddling_right) {
+                metadata.put(EntityData.PADDLE_TIME_RIGHT, 0f);
+            }
+            else {
+                paddle_time_right = 0f;
+                session.getConnector().getGeneralThreadPool().schedule(() ->
+                                updateRightPaddle(session, entityMetadata),
+                        100,
+                        TimeUnit.MILLISECONDS
+                );
+            }
+        } else if (entityMetadata.getId() == 13) {
+            // Possibly - I don't think this does anything?
+            metadata.put(EntityData.BOAT_BUBBLE_TIME, entityMetadata.getValue());
         }
 
         super.updateBedrockMetadata(entityMetadata, session);
     }
+
+    public void updateLeftPaddle(GeyserSession session, EntityMetadata entityMetadata) {
+        while (is_paddling_left) {
+            try {
+                paddle_time_left += ROWING_SPEED;
+                metadata.put(EntityData.PADDLE_TIME_LEFT, paddle_time_left);
+                super.updateBedrockMetadata(entityMetadata, session);
+                Thread.sleep(100);
```

Comment:
Should not be sleeping the thread at all here as this sleeps every other operation too.